rust/frontend/src/ty.rs

```diff
@@ -0,0 +1,163 @@
+                TVMTypeCode::$variant
+            }
+        }
+
+        impl<'a> From<&'a $type> for TVMTypeCode {
+            fn from(_arg: &$type) -> Self {
+                TVMTypeCode::$variant
+            }
+        }
+
+        impl<'a> From<&'a mut $type> for TVMTypeCode {
+            fn from(_arg: &mut $type) -> Self {
+                TVMTypeCode::$variant
+            }
+        }
+    };
+}
+
+impl_prim_type!(TVMDeviceType, kDLInt);
+impl_prim_type!(TVMContext, kTVMContext);
+impl_prim_type!(TVMType, kTVMType);
+impl_prim_type!(Function, kFuncHandle);
+impl_prim_type!(Module, kModuleHandle);
+impl_prim_type!(NDArray, kArrayHandle);
+impl_prim_type!(TVMByteArray, kBytes);
+
+/// See the [module-level documentation](../ty/index.html) for more details.
+///
+/// Wrapper around underlying TVMType
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub struct TVMType {
+    // inner fields are (code: u8, bits: u8, lanes: u16)
+    pub inner: ts::TVMType,
+}
+
+impl TVMType {
+    pub(crate) fn new(type_code: u8, bits: u8, lanes: u16) -> Self {
+        TVMType {
+            inner: ts::TVMType {
+                code: type_code,
+                bits: bits,
+                lanes: lanes,
+            },
+        }
+    }
+}
+
+impl<'a> From<&'a str> for TVMType {
+    fn from(type_str: &'a str) -> Self {
+        match type_str {
+            "int" => TVMType::new(0, 32, 1),
+            "uint" => TVMType::new(1, 32, 1),
+            "float" => TVMType::new(2, 32, 1),
+            "handle" => TVMType::new(4, 64, 1),
+            _ => panic!("Unsupported type {:?}", type_str),
+        }
+    }
+}
+
+impl Display for TVMType {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self.inner {
+            ts::TVMType {
+                code: 0,
+                bits: 32,
+                lanes: 1,
+            } => write!(f, "int"),
+            ts::TVMType {
+                code: 1,
+                bits: 32,
+                lanes: 1,
+            } => write!(f, "uint"),
+            ts::TVMType {
+                code: 2,
+                bits: 32,
+                lanes: 1,
+            } => write!(f, "float"),
+            ts::TVMType {
+                code: 4,
+                bits: 64,
+                lanes: 1,
+            } => write!(f, "handle"),
+            _ => write!(f, "unknown type"),
+        }
+    }
+}
+
+impl From<TVMType> for ts::DLDataType {
+    fn from(dtype: TVMType) -> Self {
+        dtype.inner
+    }
+}
+
+impl From<ts::DLDataType> for TVMType {
+    fn from(dtype: ts::DLDataType) -> Self {
+        Self::new(dtype.code, dtype.bits, dtype.lanes)
+    }
+}
+
+impl Deref for TVMType {
+    type Target = ts::TVMType;
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+impl DerefMut for TVMType {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+impl<'a, 'b> From<&'b TVMType> for &'a str {
+    fn from(ty: &TVMType) -> Self {
```

Comment:
shouldn't this use bitsxlanes, too?

Reply:
Perhaps `{dtype}{bits}` if lanes is 1 and `{dtype}{bits}x{lanes}` when lanes > 1. Though I like `{dtype}{bits}x{lanes}` and it makes more sense and should be applied in python and java.

Reply:
there's already an existing convention: `{dtype}{bits}x{lanes}` that's what `packed_func.h` and python and java do